httpbp/errors.go

```diff
@@ -633,3 +638,112 @@ func ErrorForCode(code int) *ErrorResponse {
 	}
 	return InternalServerError()
 }
+
+// ClientError defines the client side error constructed from an HTTP response.
+//
+// Please see ClientErrorFromResponse for more details.
+type ClientError struct {
+	Status     string
+	StatusCode int
+	RetryAfter time.Duration
+
+	AdditionalInfo string
+}
+
+func (ce ClientError) Error() string {
+	var sb strings.Builder
+	sb.WriteString("httpbp.ClientError: ")
+	if ce.Status == "" {
+		sb.WriteString("nil response")
+	} else {
+		sb.WriteString("http status ")
+		sb.WriteString(ce.Status)
+	}
+	if ce.AdditionalInfo != "" {
+		sb.WriteString(": ")
+		sb.WriteString(ce.AdditionalInfo)
+	}
+	return sb.String()
+}
+
+// RetryAfterDuration implements retrybp.RetryAfterError.
+func (ce ClientError) RetryAfterDuration() time.Duration {
+	return ce.RetryAfter
+}
+
+// ClientErrorFromResponse creates ClientError from http response.
+//
+// It returns nil error when the response code are in range of [200, 400),
+// or non-nil error otherwise (including response being nil).
+// When the returned error is non-nil,
+// it's guaranteed to be of type *ClientError.
+//
+// It does not read from resp.Body in any case,
+// even if the returned error is non-nil.
+// So it's always the caller's responsibility to read/close the body to ensure
+// the HTTP connection can be reused with keep-alive.
+//
+// The caller can (optionally) choose to read the body and feed that back into
+// ClientError.AdditionalInfo (it's always empty string when returned), e.g.:
+//
+//     resp, err := http.Get("https://www.reddit.com")
+//     if err != nil {
+//       return err
+//     }
+//     defer httpbp.DrainAndClose(resp.Body)
+//     err = httpbp.ClientErrorFromResponse(resp)
+//     if err != nil && errors.As(err, &ce) {
```

Comment:
`ce` doesn't exist yet in this example.

Reply:
:doh: I forgot to delete the `&& ...` part when splitting the if.

Reply:
Actually I moved this to testable example so it's guaranteed to be correct code :)